SSA/Projects/Holor/Holor.lean

```diff
@@ -0,0 +1,101 @@
+import SSA.Core.WellTypedFramework
+import SSA.Core.Util
+import Mathlib.Data.Holor
+import Mathlib.Algebra.Algebra.Basic
+
+#check Holor
+
+/- Holor is profunctorial, covariant in the first argument and contavariant in the second argument. -/
+
+/-- Mapping is covariant in the value -/
+def Holor.map (f : α → β) (h : Holor α ds) : Holor β ds :=
+  f ∘ h
+
+theorem Holor.map_functorial (f : α → β) (g : β → γ) (h : Holor α ds) :
+  Holor.map g (Holor.map f h) = Holor.map (g ∘ f) h :=
+  rfl
+
+/-- Reindexing is contravariant in the index -/
+def Holor.reindex (ix : HolorIndex ds₂ → HolorIndex ds₁) (h : Holor α ds₁) : Holor α ds₂ :=
+  λ ix₂ => h (ix ix₂)
+
+theorem Holor.reindex_functorial
+  (ix : HolorIndex ds₂ → HolorIndex ds₁) (iy : HolorIndex ds₃ → HolorIndex ds₂)
+  (h : Holor α ds₁) : Holor.reindex iy (Holor.reindex ix h) = Holor.reindex (ix ∘ iy) h := rfl
+
+
+/-- Pointiwse multiplication of holors. Lifts multiplicative structure into holor. -/
+def Holor.pointwise_mul [Mul α] (h₁ h₂ : Holor α ds) : Holor α ds := fun ix => (h₁ ix) * (h₂ ix)
+
+instance  [Mul α] : Mul (Holor α ds) where
+  mul := Holor.pointwise_mul
+
+theorem Holor.pointwise_mul_index [Mul α] (h₁ h₂ : Holor α ds) : 
+  (h₁ * h₂) i = h₁ i * h₂ i := rfl
+/-- Fill a holor with a constant value 'a'. -/
+def Holor.fill (a : α) : Holor α ds := fun _ => a
+
+instance [Zero α] : Zero (Holor α ds) where 
+  zero := Holor.fill Zero.zero
+
+instance [M : MulZeroClass α] : MulZeroClass (Holor α ds) where
+  zero_mul := λ h => by
+    
+    simp only [Holor.pointwise_mul_index]
+    simp[MulZeroClass.zero_mul]
+
```

Comment:
Not sure how to finish this. Lean seems to insist on replacing a `zero` with a `ofNat 0`, which means I cannot rewrite with `MulZeroClass.zero_mul`